std/experimental/logger/asynclogger.d

```diff
@@ -0,0 +1,127 @@
+module std.experimental.logger.asynclogger;
+
+import std.experimental.logger.core;
+import std.concurrency;
+
+/** The $(D AsyncLogger) processes logs asynchronously by using another $(D Logger) implementation
+as a backend. Every data logged to this $(D AsyncLogger) will be send to separate logging thread
+in order to be logged there. It can help to improve logging performance for $(D log) caller,
+for example by removing waiting for flushing logs to a file.
+Example:
+-------------
+void main()
+{
+    import std.algorithm;
+    import std.conv;
+    import std.stdio;
+    import std.datetime;
+    import std.experimental.logger;
+
+    auto fl = new FileLogger("defaultFileLogger.log", LogLevel.all);
+    void logFileDefault()
+    {
+        fl.log("test message");
+    }
+
+    auto afl = new AsyncLogger(new FileLogger("asyncFileLogger.log", LogLevel.all));
+    void logFileAsync()
+    {
+        afl.log("test message");
+    }
+
+    auto r = benchmark!(logFileDefault, logFileAsync)(100_000);
+    writeln(map!(a => to!string(to!Duration(a)))(r[]));
+}
+-------------
+*/
+final class AsyncLogger : Logger
+{
+    /// Shuntdown message
+    struct ShutdownMsg {}
+
+    private static void spawnedFunc(shared Logger ptr) @trusted
+    {
+        auto l = cast(Logger)(ptr);
+        bool isRunning = true;
+        while (isRunning)
+        {
+            receive(
+                (LogEntryBase e)
+                {
+                    LogEntry tmp = e;
+                    tmp.threadId = ownerTid;
+                    tmp.logger = l;
+                    l.forwardMsg(tmp);
+                },
+                (ShutdownMsg msg) { isRunning = false; },
+                (OwnerTerminated ex) { isRunning = false; },
+                (Variant v) { throw new Error("unhandled message"); }
+            );
+        }
+    }
+
+    /** This constructor sets a LogLevel and starts a new logging thread
+    that logs every entry asynchronously by using Logger. Because logger writes down
+    all entries into special logging thread, logger should not be used further
+    (without confidence that logger is not being used at the moment).
+    */
+    this(Logger logger, LogLevel lv = LogLevel.all) @trusted
```

Comment:
> Giving Logger proper support for shared and accepting a shared(Logger) here.

The interface should not suppose that Logger must be shared (with shared methods, etc).
> Making the constructor generic in order to create and own the logger by itself, so at least the  unsafety is encapsulated entirely within the standard library.

The problem with passing arguments for constructor will be raised in this case.
> Accept Unique!Logger instead, requiring the user to either a) guarantee uniqueness, or b) formalize the assumption of uniqueness, making it explicit for all to see.

Yes, this is exactly a problem of passing ownership to the thread. I tried to accept Unique!Logger but @MartinNowak has already mentioned about the problem.